### ISSUE TITLE
Safari Extension Fails to Detect Correct Dark Mode

### ISSUE THREAD

**Issue by zexpe** [bitwarden/clients#2460]
### Steps To Reproduce

Oftentimes the Safari Extension doesn't honour the system Dark Mode setting - the extension would be in light mode when the system is in Dark Mode, or it would be in dark mode when the system is in Light Mode...

### Expected Result

That the extension would always switch between light mode and dark mode whenever the system changes.

### Actual Result

It's often the opposite.

### Operating System

macOS

### Operating System Version

12.3

### Web Browser

Safari

### Browser Version

15.4

### Build Version

1.56.6

**Comment by eliykat** [comment]
Can you provide any specific steps we can follow to reproduce this?

One thought - do you have your OS theme set to Auto? (i.e. it automatically uses light mode during the daytime and dark mode during nighttime.) If so, I think we're missing the listeners to automatically update the popup if you still have a popup window open (e.g. popped out or sidebar) when the transition occurs.

**Comment by zexpe** [comment]
Yes, it's set to Auto. This problem isn't unique to the Bitwarden extension. Other extensions seem to suffer from this problem too. It doesn't always happen in a clearly reproducible way but it does happen very frequently - that it's in dark mode when the browser is in light mode, or it's in light mode when the browser is in dark mode.

**Comment by zexpe** [comment]
Changing the theme to Light manually makes it return to light mode, but then returning that setting to Default then makes it go back into Dark mode.

**Comment by zexpe** [comment]
Disabling the Safari Extension for Bitwarden and then enabling it again makes it return to Light mode (the system mode) with the Default setting again.

**Comment by Hinton** [comment]
@eliykat It seems we are checking the theme using a css/js trick inside the platform service, which gets called in the background page. We might need to move that check to the popup and call it on each launch instead.

@zexpe Is it possible the bitwarden application got updated which restarted the extension?

**Comment by zexpe** [comment]
There is another odd bug(?) where sometimes when I open up the Bitwarden desktop app it reinstalls the safari extension. That didn't happen today, but it did happen yesterday. It wasn't when the desktop app was updated - that was a couple of days before...

**Comment by zexpe** [comment]
I guess it might be possible that that was the first launch of the desktop app after the AppStore update was installed, but that updated refused to install unless Safari was first closed, which gave the impression it would have updated the extension? I'll try to remember to investigate that one in more detail next time an update occurs.

**Comment by zexpe** [comment]
When the system automatically turned on dark mode, the Bitwarden extension remained in light mode.
<img width="523" alt="Screenshot 2022-03-29 at 22 02 20" src="https://user-images.githubusercontent.com/9435103/160706639-005e1f65-7752-4aef-b4c2-0ddd2e632b21.png">
<img width="575" alt="Screenshot 2022-03-29 at 22 02 30" src="https://user-images.githubusercontent.com/9435103/160706657-6ba857a9-1a4d-4311-9441-028fa341e43c.png">
Disabling then re-enabling the extension fixed it again.

**Comment by eliykat** [comment]
@Hinton yes you're right, I think we should extract the theming logic into its own service (for all clients) and make sure to run it in the popup in browser.

@zexpe not sure about the update issue, please open a new issue if it happens again and we can investigate.

**Comment by djsmith85** [comment]
Thanks to @addisonbeck the theming logic has been split out with https://github.com/bitwarden/clients/pull/2943. These changes will be included in the 2022.07 release.

@zexpe Please report back, once you've had a chance to test it 2022.07

**Comment by zexpe** [comment]
It's still clearly an issue on 2022.08, which I just updated to yesterday.
<img width="283" alt="Screenshot 2022-09-02 at 10 10 25" src="https://user-images.githubusercontent.com/9435103/188106157-40a2cb9b-d700-478d-8938-9511614ffe1c.png">

**Comment by djsmith85** [comment]
@zexpe Thank you for reporting back and I'll re-open this. 

Is there any change of behaviour with 2022.08? Any more information you can provide? Does this occur more often/less often, anything else that changed? Does this also happen when not using the `Neo Noir Safari Extension`?

**Comment by zexpe** [comment]
I've done some tests just changing the system dark mode on and off manually. Bitwarden in the "default" theme does not switch with the system mode. However, as shown Neo Noir does. The first video is with Neo Noir extension disabled, demonstrating I'm in "default" theme mode, and that manually switching the extension theme works. Returning to "default" theme mode it remains in dark mode, whatever the system mode is set to. The second video shows how the Neo Noir extension correctly detects the current system mode.

https://user-images.githubusercontent.com/9435103/188204195-e3333910-d5d1-4541-b766-a9d542c23cc4.mov


https://user-images.githubusercontent.com/9435103/188204476-4ecff5cf-71bb-462a-9d09-4de359a6e22e.mov

**Comment by zexpe** [comment]
Again, if I now disable the Bitwarden extension and then enable it again it starts up in Light mode, as per the system, and still under the "Default" theme option.

**Comment by djsmith85** [comment]
Linking https://github.com/bitwarden/clients/pull/3107 here for more context. This definitely only applies to Safari and seems to be due to an upstream issue with Safari.

[Code comment from services.module](https://github.com/bitwarden/clients/blob/3a298bd9899fd3d88cce8c46329f79ca13624478/apps/browser/src/popup/services/services.module.ts#L317): 

>  // Safari doesn't properly handle the (prefers-color-scheme) media query in the popup window, it always returns light.
    // In Safari we have to use the background page instead, which comes with limitations like not dynamically changing the extension theme when the system theme is changed.

**Comment by closebot-bw** [comment]
⚠️ **Stale Issue Notice**

This issue has been automatically marked as stale due to inactivity. It will be closed in **2 weeks** (October 15, 2025) if no further activity occurs.

If this issue is still relevant and you would like to keep it open, please:
- Comment on this issue to show continued interest
- Provide any additional information or updates
- Confirm that the issue still exists in the latest version

Thank you for your contribution to this project! 🙏

**Comment by closebot-bw** [comment]
🔒 **Issue Closed Due to Inactivity**

This issue has been automatically closed due to lack of activity for an extended period. We periodically review and close inactive issues to help maintain our issue tracker and focus on current priorities.

**If this issue is still relevant:**
- Please create a new issue with updated information
- Include steps to reproduce the problem if it's a bug report
- Mention if this issue still occurs in the latest version

Thank you for your contribution to this project. Your feedback helps us improve! 🙏
